Implementing configuration file

ErwanAliasr1 opened the pull request:
As per issue #94 & #53 , we need to implement a configuration file to tune the behavior of ceph-nano regarding the expectations of users.

This PR will implement all the needed to get a configuration mecanism to override or define some tuning parameters.

As the exisiting cn is not supporting any config file, not having one must not be fatal.

Commit:
cmd: Adding configuration file support

This patch is about adding the option to add configuration files for
defining some tuning.

The configuration file could be placed in /etc/ceph or in the home
directory and should be named "cn.toml".

The tuning parameters are considered in this order :
- if no configuration file is present and if the parameter is mandatory, ceph-nano uses its internal hardcoded value.
- if there is a configuration file:
  - if there is any internal harcoded value, that's the default value
  - if the parameter is re-defined in the [default] section, this value superseed the internal hardcoded value
  - if the parameter is re-defined in the [myclustername] section,  this value superseed any other one.

Let's consider the first implementation in this commit, the size of the container memory.

The current default value is 512MB and harcoded directly in container.Resources.memory field.

The new scheme is to define its default value in setDefaultConfig() which creates a void entry in the [default] section with this 512MB value.
So mimic a user adding an entry like the following in the configuration file :

    [default]
      MemorySize=512MB

When ceph-nano needs to pass the memory size option to docker, it uses a call to getMemorySize(containerName).
This function will extract with getValueFromConfig() the current value for this containerName.
If the container doesn't define a particular value, the [default] value will be used.

This commit also add a test framework to guarantee a stable behavior when patching this part of the code.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Implement use_default

The automatic inheriting of [default] is implicit (use_default=true) but
could be disabled in some particular cases.

If use_default=false, then a specific container should not inherit from the default value of a given parameter but neither reuse all the other defined values.

This patch is about implementing how retrieving a parameter should be
done regarding the state of use_default.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Enforcing containerName for getMemorySize

The name of the container must be present for every call to getMemorySize.
The semantic of requesting a kind of implicit default container is wrong as its always about getting the memory size of a specific container.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Reading ceph.conf parameters from config file

A container can define a [<containername>.ceph.conf] section to define a list of parameters to be passed to the ceph.conf configuration file.

This commit add a getStringMapFromConfig() function to read a set of parameters at once while taking care of the use_default status.

The getCephConf() func implement the call to getStringMapFromConfig() to retrieve the "ceph.conf".

This commit implements the associates tests accordingly.

Signed-off-by: Erwan Velu <erwan@redhat.com>

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,107 @@
+		}
+	} else {
+		viper.SetConfigName("cn")         // name of config file (without extension)
+		viper.AddConfigPath("/etc/ceph/") // path to look for the config file in
```
This won't work since the `ceph.conf` lives inside the container and is not accessible from the host.

ErwanAliasr1 replied:
This is the configuration file for nano. So that file is on the host side. I was wondering where to put it. Could be /etc/cn or /etc/ceph-nano/ or any path you'd like.

leseb replied:
Ahhh, I got confused. Typically CLI programs such as cn relies on `$HOME/.cn/`

leseb replied:
Ok you're calling it multiple times... Then I believe `/etc/cn` is acceptable.

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,107 @@
+	return int64(bytes)
+}
+
+func getCephConf(containerName string) map[string]interface{} {
```
See my comment above.

Commit:
config: Implement cpu_count

By default, ceph-nano was using a single cpu per container.
This remains the default but could be overrided by specifying "cpu_count=<x>" in the container section of the configuration file.

This commit add also getInt64FromConfig() to extract an int64 from a parameter.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Reporting configuration file used

When loading the configuration file, it could come from various places.
Let's print which one is used to help user understanding what file to edit.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Report memory & cpu settings at start time

When starting a cluster, it's useful to know what is the memory/cpu tuning in place.
To get a human readable output, this patch rename getMemorySize() to getMemorySizeInBytes().

A new function getMemorySize() reports the value as its written in human format in the config file.

A typical output looks like :
    [user@host] cn cluster start test_nano_no_default
        Using /home/erwan/.cn/cn.toml as configuration file
        2018/12/07 12:50:43 Running cluster test_nano_no_default (1GB Memory / 2 CPU) ...

This commit updates the tests for getMemorySizeInBytes and create new
ones for the new version of getMemorySize()

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Adding flavor types to containers

When starting a container, the configuration file provides a set of default flavors to be used.

The new "-f" option allow the user to select one or "default" will be used.
"default" represent the builtin values of ceph-nano or the overrided ones in the [default] section of the configuration file.

A check is done at startup to ensure the requested flavor exists.

To be consistent with this vision, the containerName variables in the code are renamed to containerFlavor.

This commit also brings a potential default configuration file (cn.toml), to be adjusted over time, to offer a set of default flavors to be used.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
main: Reporting if no configuration file found

This inform users when there is an issue at loading the configuration file.
That's not a fatal error as cn can run without a config file.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Renaming MemorySize to memory_size

To keep more consistency with other variables, let's rename this parameter.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Moving flavors into [flavors]

Future work will requires having images inside the configuration file.
This patch will move everything into [flavors].

The CLI remains untouched, so there is no need to add "flavors" in the -f option.

This patch updates the API of getStringFromConfig, getStringMapFromConfig, getInt64FromConfig, useDefault to explicit the group to reach.
Today, only the "flavors" group exists. That will make this code more generic.

As per the code, the configuration file is structured like:
[group]
  [item]
    itemValue=value

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Adding [images] support

Without a configuration file, the "-i" option allow users to choose a particular container to be used.
This behavior is kept for compatibility.

When running with a configuration file, the -i option points to an entry into the configuration file.

    cn cluster start -i redhat

In this example, the configuration file have this redhat entry defined as in :
[images]
  [images.redhat]
    image_name="registry.access.redhat.com/rhceph/rhceph-3-rhel7"

The image_name variable is used as input parameter to docker to pull the right image.
This configuration option makes easier to remember/use some complex/long urls.

This option can be combined with the flavor (-f) option to select the proper size of container.
It is then possible to run :

    cn cluster start -i redhat -f huge
or
    cn cluster start -i redhat -f medium

Signed-off-by: Erwan Velu <erwan@redhat.com>

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,173 @@
+	return getStringFromConfig(FLAVORS, containerFlavor, "memory_size")
+}
+
+func getMemorySizeInBytes(containerFlavor string) int64 {
```
Shouldn't we put this in `utils.go` instead?

ErwanAliasr1 replied:
done.

leseb commented on cmd/main.go:
```diff
@@ -130,8 +133,10 @@ func Main(version string) {
-	if os.Getenv("CN_REGISTRY") == "redhat" {
-		imageName = "registry.access.redhat.com/rhceph/rhceph-3-rhel7"
+	if configurationFile = readConfigFile(); len(configurationFile) > 0 {
+		fmt.Printf("Using %s as configuration file\n", configurationFile)
```
Use `log.Println` instead so we are consistent with the existing code.

leseb replied:
It's a bit 'annoying' to see this message each time we run the CLI.
Maybe we could have an option to suppress it in the config file?

ErwanAliasr1 replied:
Printf have the advantage to ease the formating. It that critcial to have only println ?

ErwanAliasr1 replied:
Regarding my experience during the devel, as the behavior changes when you have a config file or not, and as the configuration file could be in several places, that is very difficult to anticipate if the bad behavior is due to the presence/lack/wrong config file.
That's why I put this at every start, so it's explicit. I know its verbose but useful.

ErwanAliasr1 replied:
@leseb This is the last item to handle.

leseb commented on cmd/main.go:
```diff
@@ -130,8 +133,10 @@ func Main(version string) {
+	if configurationFile = readConfigFile(); len(configurationFile) > 0 {
+		fmt.Printf("Using %s as configuration file\n", configurationFile)
+	} else {
+		fmt.Println("No configuration file found")
```
ditto

ErwanAliasr1 replied:
Solved

leseb commented on cmd/start.go:
```diff
@@ -65,17 +69,26 @@ func cliClusterStart() *cobra.Command {
+	// Ensure the flavor exists or report an error
+	if !isEntryExists(FLAVORS, flavor) {
+		if len(configurationFile) == 0 {
+			panic("Using flavors requires a configuration file")
```
Use `fmt.Errorf` to stay consistent.

ErwanAliasr1 replied:
Fixed with log.Fatal()

leseb commented on cmd/start.go:
```diff
@@ -65,17 +69,26 @@ func cliClusterStart() *cobra.Command {
+		if len(configurationFile) == 0 {
+			panic("Using flavors requires a configuration file")
+		}
+		panic("The flavor " + flavor + " doesn't exists")
```
ditto

ErwanAliasr1 replied:
Fixed with log.Fatal()

leseb commented on cmd/start.go:
```diff
@@ -65,17 +69,26 @@ func cliClusterStart() *cobra.Command {
+		if len(configurationFile) == 0 {
+			panic("Using flavors requires a configuration file")
+		}
+		panic("The flavor " + flavor + " doesn't exists")
```
s/exists/exist/

ErwanAliasr1 replied:
Fixed

leseb commented on cmd/utils.go:
```diff
@@ -914,3 +914,21 @@ func listRedHatRegistryImageTags() {
+
+	// If there is a '-i' argument, let's check if the entry exists or return an error
+	if !isEntryExists(IMAGES, imageName) {
+		panic("Image " + imageName + " doesn't exists")
```
`fmt.Errorf`

ErwanAliasr1 replied:
Fixed with log.Fatal()

leseb commented on cmd/utils.go:
```diff
@@ -914,3 +914,21 @@ func listRedHatRegistryImageTags() {
+
+	// If there is a '-i' argument, let's check if the entry exists or return an error
+	if !isEntryExists(IMAGES, imageName) {
+		panic("Image " + imageName + " doesn't exists")
```
s/exists/exist/

ErwanAliasr1 replied:
Fixed

leseb commented on cn.toml:
```diff
@@ -0,0 +1,34 @@
+title = "Default configuration file"
```
`Default Ceph Nano configuration file`

ErwanAliasr1 replied:
Fixed.

leseb commented on cmd/utils.go:
```diff
@@ -914,3 +914,21 @@ func listRedHatRegistryImageTags() {
+	}
+
+	// If there is a '-i' argument, let's check if the entry exists or return an error
+	if !isEntryExists(IMAGES, imageName) {
```
What if the entry does not exist in the config file but I want to explicitly boot an image? without editing the config file. I don't think we should fail, we should try to boot anyway, if the image does not exist the registry will let us know.

ErwanAliasr1 replied:
So if that name exists we use it from the config file unless we try directly to pull one named like that.

leseb replied:
Yes, so let's not fail.

leseb reviewed (changes requested):
A bunch of comments:

* `cn cluster ls` should expose a `FLAVOR` field too.
* it will be nice to do a `cn config show <flavor>` maybe?

Commit:
config: Switching from panic() to log.Fatal()

panic isn't the best way to report a message to the user. It generates a
backtrace which isn't suitable for such user interaction.

log.Fatal() is much more appropriate as it does cleaning print a message + os.Exit(1).

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Hardcode more flavors & images

The default configuration file was featuring most of the useful entries.
To get a common experience between already running cn instances (without configuration files) and the newest (having one) harcoding some of most used values is useful.
This also put the config file as a place where people put their custom setup only.
This ease the modification/injection of custom setups as taking care of the default values will be avoided.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Adding "flavors {ls|show}" commands

This patch is about adding two new commands on the cli.

- cn flavors ls
This command will list all the defined flavors (hardcoded + user defined).
It reports the name and the basic information (memory & cpu).
A typical output looks like :

    [erwan@host]$ cn flavors ls

    Using /home/erwan/.cn/cn.toml as configuration file
    +----------------------+-------------+-----------+
    | NAME                 | MEMORY_SIZE | CPU_COUNT |
    +----------------------+-------------+-----------+
    | huge                 | 4GB         | 2         |
    | medium               | 768MB       | 1         |
    | large                | 1GB         | 1         |
    | test_nano_no_default | 1GB         | 2         |
    | default              | 512MB       | 1         |
    +----------------------+-------------+-----------+

- cn flavors show <flavor>
This command will list the detail of a given flavor
It report the definition of it.
A typical output looks like :

    [erwan@host]$ cn flavors show test_nano_no_default

    Details of the flavor test_nano_no_default:
    {
      "ceph": {
        "conf": {
          "osd_memory_target": 3841234556
        }
      },
      "cpu_count": 2,
      "memory_size": "1GB",
      "use_default": false
    }

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Using viper.IsSet() instead of custom code

Testing the presence of a key inside a viper configuration should be
done by the builtin call (IsSet) instead of a custom code.

The isEntryExists() function is kept as it wraps the group + item usage
into it. Pretty simple function but useful for the semantic.

Adding some tests to ensure a stable behavior over time.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Moving high level functions from config_file to utils

Some high level functions are useful in many place in the code so let's
move them into utils to keep config_file only having the basic internals to manipulate the config files & defaults.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
cn.toml: Updating title

Updating title to match the project name.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
image: Adding 'image show-aliases' command

The configuration file have bring an internal representation of registries as an alias name.
This patch expose the list of possible aliases to be used on the '-i' option.

A typical output looks like :

    erwan@host: $ cn image show-aliases
    Using /home/erwan/.cn/cn.toml as configuration file
    +----------+--------------------------------------------------+
    | ALIAS    | IMAGE_NAME                                       |
    +----------+--------------------------------------------------+
    | redhat   | registry.access.redhat.com/rhceph/rhceph-3-rhel7 |
    | luminous | ceph/daemon:latest-luminous                      |
    | mimic    | ceph/daemon:latest-mimic                         |
    +----------+--------------------------------------------------+

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
start: Removing privileged option from command line

The privileged option was exposed to the user via the cli but didn't really meant to be directly used.
When needed this value was set directly by the code itself.

This patch is about to remove the option from the cli while keeping an option to enforce it in the flavors.
This could let an option to reenable it quickly if some need it.
After some time, if that is no more used, this could be totally removed.

The default value remain set to false.

This commit also add getBoolFromConfig() to retrieve a bool from the configuration.
It also add the needed tests to guarantee stability over time.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
Makefile: Add tests in the prepare target

When preparing the build, running the test could detect a potential failure.
In such case, the build process must be aborted.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Add "--data" support in flavors

When defining a flavor, it's possible to define the default value of the "--data" cli option.
The flavor can define a "data" paramater to define it.
If the use pass the '--data' on the command line, this one will be prefered over the flavor defined one.

The internal function to read this value is called getUnderlyingStorage() to keep a good semantic.
The flavor parameter is kept to "data" to mimic the option name.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Adding '--size' support in flavors

When defining a flavor, it's possible to define the default value of the "--size" cli option.
The flavor can define a "size" paramater to define it.
If the use pass the '--size' on the command line, this one will be prefered over the flavor defined one.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Adding '--work-dir' support in flavors

When defining a flavor, it's possible to define the default value of the "--work-dir" cli option.
The flavor can define a "work_directory" paramater to define it.
If the use pass the '--work-dir' on the command line, this one will be prefered over the flavor defined one.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
utils: Unifying toBytes() conversions

getMemorySizeInBytes() is using units.ParseBase2Bytes() to convert human readable units into bytes.
The existing code was using a toBytes() function to perform the same operation.

This commit is about having a single function relaying on units.ParseBase2Bytes().
This way the code is more robust & shared between functions.

This functions is tested with getMemorySizeInBytes() checks.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Using isParameterExist() instead of custom IsSet() calls

This patch is about replacing some IsSet() calls by a generic isParameterExist() which uses group, item & parameter fields.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Reworking merging of default & other flavors

When an item is defined in [default] and if a flavor keep the
"use_default=true" parameter, the item should be present in the flavor.

This patch is adding a new_param=true in the testing cn file so it
should be found in [test_nano_default] flavor and not in the [test_nano_no_default] one.

This new test proven this feature to be broken.

This patch does :
- switch use_default to a real bool and no more a string

- add mergeFlavorsWithDefault() function
  called by ReadConfigFile() to make all items of [default] being populated in other flavors when use_default is set to true.

- add getDefaultParameters() to list all parameters from [default]
  The default parameters aren't reachable from a Get() call, so a workaround is needed to get them.

- remove all the default parameter extraction from get*FromConfig()
  This makes the code simplier & easier to update

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Populating 'flavors show default' command

This command was broken and didn't reported anything as all the items were set with SetDefault().

This patch is about calling getDefaultParameters() when asking 'show default'.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Adding more parallelism in tests

The tests were written to be run a single time but it appear that running some of them several time broke the suite.

One of the reason is that once the ReadConfigFile() is done, the context is changed and doesn't really represent a default status.

This commit is about :
- having a prepare target that tries to run the test 5 times
- split all the tests on the default config to be run first
- split all the tests on the config file after a single read of the config file
- ensure that if some internal variables are used to triggering some particular behaviors, variables are restored after the test

This patch made the tests being more robust, it is now possible to run them 5 times without any issue.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Fixing typo

s/exists/exist/

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Don't print a message if no configuration file found

As not having a configuration file is not fatal, that's pretty verbose to report that no configuration file got found.

Signed-off-by: Erwan Velu <erwan@redhat.com>

ErwanAliasr1 commented:
@leseb This PR is over for me.

ErwanAliasr1 commented:
once merged, that could make a 2.1 release.

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+	"github.com/spf13/viper"
+)
+
+//FLAVORS is a constant to represent the [flavors] group
```
```suggestion
// FLAVORS is a constant to represent the [flavors] group
```

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+//FLAVORS is a constant to represent the [flavors] group
+const FLAVORS = "flavors"
+
+//IMAGES is a constant to represent the [images] group
```
```suggestion
// IMAGES is a constant to represent the [images] group
```

Commit:
Packaging: Adding cn.toml configuration file

This commit is about adding the sample configuration file as default configuration file in /etc/cn/

Signed-off-by: Erwan Velu <erwan@redhat.com>

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+			log.Fatal(err)
+		}
+
+		// Let's import all the default value into flavors
```
```suggestion
		// Let's import all the default values into flavors
```

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+			continue
+		}
+
+		// Nothing to do if the flavor set the use_default=false
```
```suggestion
		// Nothing to do if the flavor sets the use_default=false
```

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+// Set the default values for defined types
+// If the configuration file is missing, this section will generated the mandatory elements
```
```suggestion
// If the configuration file is missing, this section will generate the mandatory elements
```

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+func getStringFromConfig(group string, item string, name string) string {
+	var value = ""
```
```suggestion
	var value string
```

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+	var value = ""
+	// If we are requested to get the status of use_default, we cannot call useDefault ;)
+	itemValue := viper.GetString(group + "." + item + "." + name)
+	if len(itemValue) > 0 {
```
What if we don't enter this condition? We return an empty string, shouldn't we error?

ErwanAliasr1 replied:
Comment was wrong, code was old. I reworked that part. Thx.

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	// If we are reaching this point, let's check if we are in the chicken/egg case trigger by mergeFlavorsWithDefault
+	// As mergeFlavorsWithDefault checks if use_default is while its not yet populated
```
Not sure to understand what you meant here

leseb commented on cmd/config_file.go:
```diff
@@ -0,0 +1,232 @@
+	return viper.IsSet(group + "." + item + "." + parameter)
+}
+
+// Afunction to list the default parameters as they are not seen
```
```suggestion
// A function to list the default parameters as they are not seen
```

leseb commented on cmd/flavors.go:
```diff
@@ -0,0 +1,91 @@
+// cliFlavorsShow is the Cobra CLI call
+func cliFlavorsShow() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "show",
```
```suggestion
		Use:   "show [flavor]",
```

leseb commented on cmd/flavors.go:
```diff
@@ -0,0 +1,91 @@
+func cliFlavorsShow() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "show",
+		Short: "Show a flavor",
```
Add:

`		DisableFlagsInUseLine: true,`

leseb commented on cmd/flavors.go:
```diff
@@ -0,0 +1,91 @@
+	flavorName := args[0]
+	flavor := FLAVORS + "." + flavorName
+	if isEntryExist(FLAVORS, flavorName) {
+		fmt.Println("\nDetails of the flavor " + flavorName + ":")
```
Why `\n`?

leseb commented on cmd/image.go:
```diff
@@ -38,5 +41,29 @@ func init() {
+		Use:   "show-aliases",
+		Short: "Print the list of aliases",
+		Args:  cobra.NoArgs,
+		Run:   listAliases,
```
Add: 		`DisableFlagsInUseLine: true,`

leseb commented on cmd/start.go:
```diff
@@ -57,35 +57,43 @@ func cliClusterStart() *cobra.Command {
+	// Ensure the flavor exist or report an error
+	if !isEntryExist(FLAVORS, flavor) {
+		if len(configurationFile) == 0 {
+			log.Fatal("Using flavors requires a configuration file")
```
I don't think this is true anymore, given that flavors are internal now.

leseb commented on cmd/utils.go:
```diff
@@ -914,3 +897,95 @@ func listRedHatRegistryImageTags() {
+	return getStringFromConfig(FLAVORS, containerFlavor, "memory_size")
+}
+
+// getMemorySizeInBytes transform a user-defined input (like 1GB) in bytes
```
```suggestion
// getMemorySizeInBytes transforms a user-defined input (like 1GB) in bytes
```

leseb commented on cmd/utils.go:
```diff
@@ -914,3 +897,95 @@ func listRedHatRegistryImageTags() {
+	return toBytes(getMemorySize(containerFlavor))
+}
+
+// getCPUCount return the number of CPUs for a flavor
```
```suggestion
// getCPUCount returns the number of CPUs for a flavor
```

leseb commented on cmd/utils.go:
```diff
@@ -914,3 +897,95 @@ func listRedHatRegistryImageTags() {
+	return getInt64FromConfig(FLAVORS, containerFlavor, "cpu_count")
+}
+
+//getCephConf return the Ceph configuration for a flavor
```
```suggestion
//getCephConf returns the Ceph configuration for a flavor
```

leseb reviewed (changes requested):
We need to update bash_completion too. Thanks!

Commit:
config: Adding documentation

This commit is about adding CONFIGURATION.md that expose the
configuration options while considering builtins, configuration files,
flavor and image aliases.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
utils: Using switch() statement in inspectImage()

Having a list of if() {return} isn't really nice to read.
switch() is really made for that.

Making the code using switch() for a better readability but also to make further contributions easier.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
list: Reporting cluster flavor

When listing clusters, it could be very useful to report the selected flavor at start time.
To do so, dockerInspect() must read the Labels from the running container.
The showNanoClusters then, have to report this value to the table displayed to the user.

This patch is then adding the "flavor" support argument to dockerInspect() but also refactor it to use the switch() statement.
Using switch() statement makes the code easier to read & to expand.

If not flavor is found, already running containers, the "none" value is printed.

A typical output looks like :

$ ./ cn cluster ls
+---------------+---------+------------------------------+----------------+--------------------------------+--------+
| NAME          | STATUS  | IMAGE                        | IMAGE RELEASE  | IMAGE CREATION TIME            | FLAVOR |
+---------------+---------+------------------------------+----------------+--------------------------------+--------+
| monpetitponey | running | docker.io/ceph/daemon:latest | master-7150c5c | 2018-08-31T12:40:14.676023593Z | medium |
+---------------+---------+------------------------------+----------------+--------------------------------+--------+

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
quality: Improving english & style

This commit is about fixing typos, style and mistakes after a proof-read from Sebastien Han.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Update getStringFromConfig() to mimic other get*FromConfig()

This function was having a wrong comment since use_default is now a
bool.
The code was also not up to date as it's better checking the existence
of the entry with isParameterExist() rather than testing the length of
the resulting string.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
image: Split cliShowAliases & listAliases

This commit is about moving cliShowAliases & listAliases functions to
image_aliases.go to make this part more look like the rest of the
project.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
flavors: Improving help message

This commit makes the user experience a little bit better.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
start: Removing wrong statement

The flavors are no more related to configuration files.

Signed-off-by: Erwan Velu <erwan@redhat.com>

leseb commented on CONFIGURATION.md:
```diff
@@ -0,0 +1,149 @@
+# Using cluster flavors
+Ceph nano offer a set of pre-defined flavors to tune your cluster :
```
```suggestion
Ceph nano offers a set of pre-defined flavors to tune your cluster:
```

leseb commented on CONFIGURATION.md:
````diff
@@ -0,0 +1,149 @@
+The flavor list can be obtained by running the `flavors ls` command:
+
+```
+$ ./cn flavors ls
````
```suggestion
$ cn flavors ls
```

leseb commented on CONFIGURATION.md:
````diff
@@ -0,0 +1,149 @@
+A flavor can also define some other tuning parameters. They can be listed by using the `flavors show` command:
+
+```
+$ ./cn flavors show default
````
```suggestion
$ cn flavors show default
```

leseb commented on CONFIGURATION.md:
````diff
@@ -0,0 +1,149 @@
+## Using a flavors
+Flavors can be selected with the `-f` flags when starting a new cluster.
+```
+$ ./cn cluster start mycluster -f huge
````
```suggestion
$ cn cluster start mycluster -f huge
```

leseb commented on CONFIGURATION.md:
```diff
@@ -0,0 +1,149 @@
+
+
+## Configuration items of a flavor
+A flavor can be tuned with various built-in items as defined below :
```
```suggestion
A flavor can be tuned with various built-in items as defined below:
```

leseb commented on CONFIGURATION.md:
```diff
@@ -0,0 +1,149 @@
+|privileged   | Defines if the container runs in privileged mode  |   false | none  |
+| use_default   | Defines if this flavor inherit from the `default` flavor  | true  | none  |
+
+If a flavors defines a `ceph.conf` sub entry, this one will be used as items for the ceph.conf configuration as per bellow :
```
```suggestion
If a flavor defines a `ceph.conf` sub entry, this one will be used as items for the `ceph.conf` configuration as per bellow:
```

leseb commented on CONFIGURATION.md:
```diff
@@ -0,0 +1,149 @@
+|image_name | Defines the associated image name for an alias| ceph/daemon|
+
+
+The following example create an alias name `sharktopus` in the `/etc/cn/cn.toml` configuration file.
```
```suggestion
The following example creates an alias name `sharktopus` in the `/etc/cn/cn.toml` configuration file.
```

leseb commented on CONFIGURATION.md:
````diff
@@ -0,0 +1,149 @@
+```
+
+# Configuration file
+Ceph nano can read its configuration from 3 different locations, they are search in the following order :
````
```suggestion
Ceph nano can read its configuration from 3 different locations, they are search in the following order:
```

leseb commented on CONFIGURATION.md:
```diff
@@ -0,0 +1,149 @@
+
+If a configuration file is found, additional flavors and images aliases are loaded from it.
+
+By default, `use_default` feature is enabled meaning that all user-defined but also built-in flavors inherit for every feature of the `default` flavor unless its defined in the new flavor.
```
```suggestion
By default, `use_default` feature is enabled meaning that all user-defined but also built-in flavors inherit for every feature of the `default` flavor unless it's defined in the new flavor.
```

leseb commented on CONFIGURATION.md:
```diff
@@ -0,0 +1,149 @@
+If `use_default` is set to false, the user-defined flavor will have not inheritance from `default` meaning all the definition of this flavor is done in the flavor section.
+
+
+In the following example, the `test1` flavor will have onl have `cpu_count` set to 2 while the `test2` flavor will have both `cpu_count` set to 2 and `memory_size` set to 640MB.
```
```suggestion
In the following example, the `test1` flavor will only have `cpu_count` set to 2 while the `test2` flavor will have both `cpu_count` set to 2 and `memory_size` set to 640MB.
```

Commit:
contrib: Improving bash_completion generation

Cobra offer a way to generate the bash_completion script.

This commit implement a new option in ceph nano (completion) to generates this script.
Calling 'cn completion' will render the script file to stdout.

This commit call this mecanism from the travis script so every release will update this file.

The spec file is updated to feature this file in the package.

The completion file is renamed to cn_completion.sh to have a more explicit name when located in /etc/bash_completion.d/

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
main: Print configuration filename to stderr

When using the output of ceph nano, like in completion, or any other
parsing, printing the configuration file name to stdout is noisy.

This patch moves this message to stderr to avoid this effect.

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
config: Merge flavors even if no configuration file found

Running a "cn flavors ls" was reporting the following message :
    cpu_count int64 value in mediumdoesn't exist

The bug was about handling setups where no configuration file got found.
ReadConfigFile() was only calling mergeFlavorsWithDefault() if a configuration file got found.
A visual artifact was also present by glueing the flavor name & the last part of the sentence.

This patch
- rework readConfigFile() to get a common code path that execute 'mergeFlavorsWithDefault()' in all cases.
- add spaces after the name of flavor.
- add a test to catch this case
- ensure that cpu_count is typed as int64 in the builtins values

Signed-off-by: Erwan Velu <erwan@redhat.com>

Commit:
flavors: Improving 'flavors show' output

When reporting the details of a flavor, it could be better and easier to
import into a 3rd party tool to only print the content of the flavor.

The introduction message was removed.
The flavor name feature a 'name' field which is automatically feed with the flavor name.
A test is added accordingly to check this feature is working properly.

Signed-off-by: Erwan Velu <erwan@redhat.com>

leseb commented on cmd/utils.go:
```diff
@@ -471,43 +473,44 @@ func dockerInspect(containerName string, pattern string) string {
+		if len(flavor) > 0 {
+			return flavor
+		}
+		return "none"
```
Unknown?
